feat(tools): custom keybinds and layer independent tool keybinds

chrisj opened the pull request:
<!-- Firebase hosting preview start -->
----
🔥 Preview 🔥: https://neuroglancer--pr464-jqtohdii.web.app
Updated at Thu Feb 22 2024 17:27:49 GMT+0000 (Coordinated Universal Time) for commit bc4f166a9f852f529781d72ce2925afd2da6a7dd
Expires: Invalid Date

<!-- Firebase hosting preview end -->

jbms commented on src/neuroglancer/ui/default_viewer_setup.ts:
```diff
@@ -20,16 +20,84 @@ import {setDefaultInputEventBindings} from 'neuroglancer/ui/default_input_event_
 declare var NEUROGLANCER_DEFAULT_STATE_FRAGMENT: string|undefined;
 
+type CustomBinding = {
+  layer: string, tool: string, protocol?: string,
```
`tool` should be a json object, not just a string, as some tools support options

jbms commented on src/neuroglancer/ui/default_viewer_setup.ts:
```diff
@@ -20,16 +20,84 @@ import {setDefaultInputEventBindings} from 'neuroglancer/ui/default_input_event_
+    });
+  }
+
+  const nameToLayer: {[key: string]: UserLayerConstructor|undefined} = {};
```
This already exists as `layerTypes` in `neuroglancer/layer.ts`

jakobtroidl commented:
@jbms, is it possible to get this merged? We need this feature in ng for proofreading H01 with CAVE.

chrisj committed:
feat(tools): custom keybinds and layer independent tool keybinds

chrisj commented:
Updated this pull request with the ability to unbind existing keybinds with a false value

Example config

```json
{
  "keym": {
    "layer": "segmentation",
    "tool": "grapheneMergeSegments",
    "provider": "graphene"
  },
  "keyc": {
    "layer": "segmentation",
    "tool": "grapheneMulticutSegments",
    "provider": "graphene"
  },
  "keyf": {
    "layer": "segmentation",
    "tool": "grapheneFindPath",
    "provider": "graphene"
  },
  "keyx": false
}
```

jbms commented on src/ui/default_viewer_setup.ts:
```diff
@@ -14,26 +14,124 @@
+  [key: string]: CustomToolBinding | string | boolean;
+};
+
+declare const CUSTOM_BINDINGS: CustomBindings | undefined;
```
Please rename to NEUROGLANCER_CUSTOM_INPUT_BINDINGS

jbms commented on src/ui/default_viewer_setup.ts:
```diff
@@ -14,26 +14,124 @@
+  if (hasCustomBindings) {
+    for (const [key, val] of Object.entries(CUSTOM_BINDINGS!)) {
+      if (typeof val === "string") {
+        viewer.inputEventBindings.global.set(key, val);
```
Note that in addition to the global map there are several others that are relevant in default_input_event_bindings.ts --- in particular the default slice view and perspective view bindings.  Perhaps this should account for that?

chrisj replied:
I added a "context" property, one of `"global" | "perspectiveView" | "sliceView"`. I could also call it map or actionMap. Restructured the json object to support it.


```json
{
  "keym": {
    "action": {
      "layerType": "segmentation",
      "tool": "grapheneMergeSegments",
      "provider": "graphene"
    }
  },
  "keyx": {
    "action": false
  },
  "control+shift+keyx": {
    "action": "clear-segments",
    "context": "sliceView"
  }
}
```

jbms commented on src/ui/default_viewer_setup.ts:
```diff
@@ -14,26 +14,124 @@
+              const protocol = viewer.dataSourceProvider.getProvider(
+                dataSource.spec.url,
+              )[2];
+              if (protocol === desiredProvider) {
```
With the kvstore changes getProvider is not exactly the right method anymore.  Perhaps just have a regexp that matches on the URL?

chrisj replied:
I switched it to use the schemePattern from `"#src/kvstore/url.js";` does that seem fine?

jbms commented on src/ui/tool.ts:
```diff
@@ -318,15 +318,15 @@ export class GlobalToolBinder extends RefCounted {
-  activate(key: string): Borrowed<Tool> | undefined {
-    const tool = this.get(key);
+  activate(key: string, tool?: Tool<object>): Borrowed<Tool> | undefined {
+    tool = tool || this.get(key);
```
Please use `tool ?? this.get(key)` instead --- not necessary here but avoids thinking about javascript truthyness complications

jbms commented on src/ui/tool.ts:
```diff
@@ -318,15 +318,15 @@ export class GlobalToolBinder extends RefCounted {
     this.debounceDeactivate.cancel();
     const activeTool = this.activeTool_;
-    if (tool === activeTool?.tool) {
+    if (tool.toJSON() === activeTool?.tool.toJSON()) {
```
We can't compare the json like this because it can be an object rather than a plain string.  Instead we need to compare the serialized json.  However, it would be better if we can avoid having to insert the json comparison here, and instead just ensure that the same `tool` object is passed.  Is your logic involving `previousTool` not sufficient?

chrisj replied:
I think originally my previousTool was for deactivation but then I realized if you have a custom binding and an user specified binding for the same tool+layer. It is a bit awkward. It is not a big deal because very few users would swap between both so this code can be removed but I made it so they can deactivate each other. For example if `c` is bound to multicut tool and the user binds shift+a to multicut, they can activate it with either bind and deactivate it with the other.

```
activeTool.tool.constructor === tool.constructor &&
activeTool.tool.context === tool.context
```

I think this logic is cleaner, let me know. If you prefer, I'm fine with removing it. Also because of this logic, previousTool has less value, really just a cache for `restoreTool`.   I could remove it and call `restoreTool` with every keypress.

jbms commented on src/ui/default_viewer_setup.ts:
```diff
@@ -14,26 +14,124 @@
 declare let NEUROGLANCER_DEFAULT_STATE_FRAGMENT: string | undefined;
 
+type CustomToolBinding = {
+  layer: string;
```
rename to `layerType` for consistency with the tool palette queries.

chrisj committed:
custom keybinds now supports modifying the other actionMap contexts (perspectiveView, sliceView)
cleaned up code, addressed PR comments

chrisj committed:
custom keybinds now optionally takes an array of bindings per identifier, allowing multiple contexts to be modified

chrisj committed:
fix setCustomInputEventBindings now only unbinds (supresses) identifiers for the target context

chrisj committed:
only accept custom tool binds for unmodified letter keys

chrisj committed:
add documentation for CustomInputEventBindings

chrisj commented:
Example config:
```json
{
  "keyh": {
    "action": "help"
  },
  "backslash": {
    "context": "global",
    "action": false
  },
  "at:control+wheel": {
    "context": "sliceView",
    "action": {
      "action": "zoom-via-wheel",
      "preventDefault": true
    }
  },
  "at:shift+mousedown0": [
    {
      "context": "sliceView",
      "action": {
        "action": "rotate-via-mouse-drag",
        "stopPropagation": true
      }
    },
    {
      "context": "perspectiveView",
      "action": {
        "action": "translate-via-mouse-drag",
        "stopPropagation": true
      }
    }
  ],
  "keyq": {
    "action": {
      "layerType": "segmentation",
      "tool": "selectSegments"
    }
  },
  "keyw": {
    "action": {
      "layerType": "segmentation",
      "provider": "precomputed",
      "tool": {
        "type": "selectSegments"
      }
    }
  }
}
```

chrisj commented:
I added documentation here: https://neuroglancer-docs--pr464-wl93s40n.web.app/json/api/index.html#configuration

I'm not sure if these json config files should go in the JSON API tab. I find it is a little bit harder to follow than just looking at an example as above but it is important to understand the event_action_map and the EventIdentifier schema.

I fixed a few edge cases and locked down tool keybinds to single letter for now as that is what ToolBinder will currently support.
